Hw/hw3

KyMuC opened the pull request:
Документация API в requests.md в корне проекта.

KyMuC committed:
Delete unnecessary configs

(cherry picked from commit 6fac30536259f3e1a4c061354f637b5e4927bdad)

KyMuC committed:
Add REST API
Add DTOs
Implement controllers
Add validators
Implement REST service layer
Update JavaDoc
Add API documentation

ilya-sukhachev commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,6 @@
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.1-all.zip
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.1-bin.zip
```
gradle-6.7

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+ * @author Mikhail_Kudimov
+ */
+@RestController
+@RequestMapping("/api/album")
```
@RequestMapping("/api/v1/album")

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+     * @param name Name to find albums by
+     * @return Albums with specified name or empty album with an error if no albums were found
+     */
+    @GetMapping("/info")
```
@GetMapping("{id}"), никаких постфиксов в виде info

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+     * @return Albums with specified name or empty album with an error if no albums were found
+     */
+    @GetMapping("/info")
+    public Collection<AlbumDto> getAlbums(@RequestParam("q") String name) {
```
RequestParam убери

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+     * @param body JSON body of album to create
+     * @return The newly created album or album with error if JSON format is wrong
+     */
+    @PostMapping("/add")
```
add не нужно

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+     * @return JSON with updated fields or JSON with fields to update and error if JSON was wrong
+     */
+    @PutMapping("/update")
+    public AlbumDto updateAlbum(@RequestParam("id") UUID id, @RequestBody AlbumDto body, BindingResult result,
```
в чате скидывал как проектировать API, не каких RequestParam. @PutMapping("{id}")

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+     * @param id Id to delete album by
+     * @return Empty album with HTTP status of 204 or album with error if album was not found
+     */
+    @DeleteMapping("/remove")
```
@DeleteMapping("{id}")

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+            return body;
+        }
+
+        if (!restAlbumService.updateAlbum(id, body)) {
```
зачем id передавать внутрь? body также же содержит id

KyMuC replied:
Тело PUT запроса не предполагает наличие идентификатора внутри него, а его передачу через PathVariable, насколько я понимаю.
Или имеется в виду, что нужно сделать body.setId(id) и в имплементации updateAlbum этот id снова достать?

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,132 @@
+
+        return albumDto;
+    }
+
```
Аналогично и в остальных контроллерах исправить.

ilya-sukhachev commented on src/main/java/ru/iteco/service/rest/RestAlbumServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+
+        albumService.addAlbum(album);
+
+        Artist artist = artistService.findById(albumDto.getArtistId());
```
почему делаешь запрос artistService? вызывай напрямую dao, аналогично и в остальных местах

KyMuC committed:
Update controllers' URLs
REST services now use DAO objects instead of model services
Update API documentation

KyMuC committed:
Pass id property through body when updating entities

KyMuC committed:
Update PUT methods' id comparison

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,139 @@
+            return body;
+        }
+
+        if (Objects.equals(id, body.getId())) {
```
избыточная вложенность if

if (!Objects.equals(id, body.getId())) {
            body.setErrors(Collections.singletonList(new ObjectError("albumDto",
                    new String[]{"id.wrong"}, null,
                    "Specified id should match body id.")));
}

if (!restAlbumService.updateAlbum(body)) {
                body.setErrors(Collections.singletonList(new ObjectError("albumDto",
                        new String[]{"album.notFound"}, null,
                        "No album found with such id.")));
            }

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,139 @@
+     * @return Albums with specified name or empty album with an error if no albums were found
+     */
+    @GetMapping("{name}")
+    public Collection<AlbumDto> getAlbums(@PathVariable String name) {
```
нужен  getById

ilya-sukhachev commented on src/main/java/ru/iteco/controller/AlbumController.java:
```diff
@@ -0,0 +1,139 @@
+            return body;
+        }
+
+        return restAlbumService.addAlbum(body);
```
статус 201. и id  в заголовке
